test/archethic/contracts/interpreter/scope_test.exs

```diff
@@ -5,55 +5,589 @@ defmodule Archethic.Contracts.Interpreter.ScopeTest do
+
+      assert %{
+               "context_list" => ["context_2"],
+               "var1" => %{"prop" => 1},
+               "var2" => 4,
+               "context_2" => %{
+                 "scope_hierarchy" => []
+               }
+             } = Process.get(:scope)
+    end
+  end
+
+  describe "create/0" do
+    test "should create scope in current context" do
+      scope = %{
+        "context_list" => ["current_context", "other_context"],
+        "var1" => %{"prop" => 1},
+        "var2" => 4,
+        "current_context" => %{
+          "scope_hierarchy" => []
+        },
+        "other_context" => %{
+          "scope_hierarchy" => []
+        }
+      }
+
+      Process.put(:scope, scope)
+
+      Scope.create()
+
+      %{
+        "current_context" => %{
+          "scope_hierarchy" => [scope_ref]
+        }
+      } = Process.get(:scope)
+
+      assert %{
+               "context_list" => ["current_context", "other_context"],
+               "var1" => %{"prop" => 1},
+               "var2" => 4,
+               "current_context" => %{
+                 "scope_hierarchy" => [^scope_ref],
+                 ^scope_ref => %{}
+               },
+               "other_context" => %{
+                 "scope_hierarchy" => []
+               }
+             } = Process.get(:scope)
+    end
+
+    test "should add another scope to an existing context hierarchy, inside current scope" do
+      # Setup with an existing scope in the current context
+      scope = %{
+        "context_list" => ["current_context"],
+        "current_context" => %{
+          "scope_hierarchy" => ["first_scope", "second_scope"],
+          "first_scope" => %{
+            "second_scope" => %{}
+          }
+        }
+      }
+
+      Process.put(:scope, scope)
+
+      Scope.create()
+
+      %{
+        "current_context" => %{
+          "scope_hierarchy" => ["first_scope", "second_scope", new_scope_ref]
+        }
+      } = Process.get(:scope)
+
+      assert %{
+               "context_list" => ["current_context"],
+               "current_context" => %{
+                 "scope_hierarchy" => ["first_scope", "second_scope", ^new_scope_ref],
+                 "first_scope" => %{
+                   "second_scope" => %{
+                     ^new_scope_ref => %{}
+                   }
+                 }
+               }
+             } = Process.get(:scope)
+    end
   end
 
-  test "write_cascade/3" do
-    Scope.init()
+  describe "leave_scope/0" do
+    test "should remove the last scope from the scope_hierarchy and delete the corresponding map" do
+      # Setup with a context that has two scopes
+      scope = %{
+        "context_list" => ["current_context"],
+        "current_context" => %{
+          "scope_hierarchy" => ["first_scope", "second_scope"],
+          "first_scope" => %{
+            "second_scope" => %{}
+          }
+        }
+      }
+
+      Process.put(:scope, scope)
+
+      Scope.leave_scope()
+
+      assert %{
```

Comment:
Here the test is not fully right as the function `leave_scope` does not delete the scope in the map, but remove only the ref in scope hierarchy.
The test is passing because you pattern match the result. Maybe you should not pattern match but use a strict equal `==` so you ensure the map is the expected one without any other data inside.
You could do it for all tests 

(Tested with `==` test is not passing)